### PR TITLE
Split DataContainer to a HDF and non HDF part

### PR DESCRIPTION
**Pull request by pmrv**
Basically DataContainer is renamed to DataContainerBase and the methods used for HasHDF/HDFStub are moved to a new sub class DataContainer. Docstrings are replicated where necessary.  No API changes for users.

In preparation of https://github.com/orgs/pyiron/projects/7/views/1

@liamhuber Pulling `DataContainer` out should be just a matter of copy/paste now.

### CODE REVIEW

**Comment by liamhuber** [comment]
> @liamhuber Pulling DataContainer out should be just a matter of copy/paste now.

I think this may be of even more imminent utility to @samwaseda?

Will review shortly, but I peeked at the diff and it was what I intuitively expected so that should go pretty smoothly.

File: pyiron_base/storage/datacontainer.py
Comment (liamhuber) [inline pyiron_base/storage/datacontainer.py:50]: If we move this part off to its own repo (e.g. `pyiron/snippets`) and leave the child class here, then in the other location I would simply call this `DataContainer` and here do like `from pyiron.snippets import DataContainer as DataContainerBase`. In the context of making the break-apart go as smoothly as possible, I am totally behind the naming scheme used in this PR, I just want to double check that we have the same middle-term vision on naming here.

File: pyiron_base/storage/datacontainer.py
Comment (liamhuber) [inline pyiron_base/storage/datacontainer.py]: Is this (and the other nearby `type(self)` appearances) really correct? I naively would have thought this should be `isinstance(v, DataContainerBase))`, since that is already what guarantees us to have `.to_builtin`. 

What _should_ the behaviour be if a `DataContainer` instance has a `DataContainerBase` instance as an item? I would want it to "just work", but I know that is vague. Here I think it certainly wouldn't as the type check would evaluate `False` and we'd get kicked over to the `repr` branch.

I don't know what the exact right behaviour is, but I think this needs a test added to solidify whatever choice you make.

File: pyiron_base/storage/datacontainer.py
Comment (liamhuber) [inline pyiron_base/storage/datacontainer.py]: Similar... if `DataClass` hits a `DataClassBase` instance, do we want to stop or not? Because right now it's "not". I could be convinced that's the right behaviour, but intuitively it feels wrong.

File: pyiron_base/storage/datacontainer.py
Comment (liamhuber) [inline pyiron_base/storage/datacontainer.py]: Similarly, should this be `(v, DataContainer)`? If my `MyDataContainer(DataContainer)` has `DataContainer` instances as items, don't those instances contain enough power to recurse like this?

**Review by liamhuber — CHANGES_REQUESTED** [review]
99% as expected and no nits or anything.

The only thing I'm not sure of is the consistent use of `type(self)` for a bunch of recursion checks. Most of these seem to me like they should instead hard-code the class they're written in, so that if derived classes hold an instance of their parent class the recursion can continue uninterrupted. I'm not totally sure that's the right behaviour choice though.

At any rate, now that we explicitly have a base and child class, I think it's important to extend the tests to cover the case of child-holds-parent-instance to encode whatever behaviour decision you make.

File: pyiron_base/storage/datacontainer.py
Comment (pmrv) [reply pyiron_base/storage/datacontainer.py:50]: Yeah, that's exactly how I thought it would go down.

File: pyiron_base/storage/datacontainer.py
Comment (samwaseda) [inline pyiron_base/storage/datacontainer.py:955]: Does it have a particular reason that it assigns `value` here?

Comment (pmrv) [reply pyiron_base/storage/datacontainer.py:955]: Once we have loaded a stub, we don't want to load it again a second time, so this overwrites the stub with the actual value.  I guess

```python
return self[key] := value.load()
```

would work as well, but is only available in newer python version (not sure which atm).

Comment (samwaseda) [reply pyiron_base/storage/datacontainer.py:955]: ok it's a small thing, but my problem is that by returning `value` it might give the impression that it's the copy of the data, while in reality if you change it, it changes the content in `self[key]` as well. So I would simply set `self[key] = value.load()` and `return self[key]`.

Comment (pmrv) [reply pyiron_base/storage/datacontainer.py:955]: Won't work because we are inside `__getitem__`, so using `self[key]` will recurse without stopping.

Comment (samwaseda) [reply pyiron_base/storage/datacontainer.py:955]: ah ok. Beautiful

Comment (pmrv) [inline pyiron_base/storage/datacontainer.py:955]:
```suggestion
            return self[key] := value.load()
```

the "walrus" is already in since 3.8 it turns out, so let's just use it.

Comment (pmrv) [reply pyiron_base/storage/datacontainer.py:955]: Ah, bummer it's not allowed in this place.  Will revert.

**Review by niklassiemer — COMMENTED** [review]
LGTM once @liamhuber s concerns are matched

File: pyiron_base/storage/datacontainer.py
Comment (pmrv) [reply pyiron_base/storage/datacontainer.py]: Agree with your reasoning, it seems a more useful choice, I just hadn't considered it.  I've changed it back, but tests will have to come later.

**Comment by pmrv** [comment]
```
 Traceback (most recent call last):
  File "/home/runner/work/pyiron_base/pyiron_base/tests/flex/test_pythonfunctioncontainer.py", line 63, in test_with_executor_wait
    self.assertTrue(job.server.future.done())
AssertionError: False is not true
```

@jan-janssen Sounds related to your recent developments.  Would that be fixed by #1300 or #1301?

**Comment by jan-janssen** [comment]
> @jan-janssen Sounds related to your recent developments. Would that be fixed by #1300 or #1301?

The issue was solved by restarting the CI. We have to take a look at this and maybe increase the time out in case the issue continues to come up.

**Comment by jan-janssen** [comment]
Once this is merged, we should also release a new `pyiron_base` version.

**Comment by pmrv** [comment]
> > @jan-janssen Sounds related to your recent developments. Would that be fixed by #1300 or #1301?
> 
> The issue was solved by restarting the CI. We have to take a look at this and maybe increase the time out in case the issue continues to come up.

I think it makes more sense to just use the timeout directly in the tests, then this can never happen.

**Comment by jan-janssen** [comment]
@pmrv As more tests were requested, I changed the status of this pull request to `draft` until the tests are added.

**Review by liamhuber — APPROVED** [review]
Nice, type checking and testing changes lgtm.

**Pull request merged**
